pallets/migration/src/lib.rs

```diff
@@ -0,0 +1,249 @@
+//! # Fees pallet for runtime
+//!
+//! This pallet provides functionality for setting and getting fees associated with an Hash key..
+//! Fees are set by FeeOrigin or RootOrigin
+#![cfg_attr(not(feature = "std"), no_std)]
+use frame_support::weights::Weight;
+
+pub use pallet::*;
+pub mod weights;
+use crate::traits::RuntimeUpgradeProvider;
+use frame_support::sp_runtime::traits::Zero;
+pub use weights::*;
+
+pub mod traits {
+	use codec::FullCodec;
+	use frame_support::dispatch::Weight;
+	use sp_version::RuntimeVersion;
+
+	pub type UsableWeight = Weight;
+	pub type UsedWeight = Weight;
+
+	// pub type Upgrader = Box<dyn FnOnce(UsableWeight) -> (Finished, UsedWeight)>;
+
+	pub trait RuntimeUpgradeProvider {
+		/// A memo to a specific upgrade. In most cases this should briefly! describe
+		/// what was done in this upgrade.
+		type Memo: Clone + Eq + sp_std::fmt::Debug + AsRef<[u8]>;
+
+		/// An info that needs to be defined by the implementer of this trait.
+		///
+		/// The calley can uses this info in order to memorize himself, which part of the
+		/// upgrade needs to be run next. The lastly received `StateInfo` will always be provided
+		/// to the implementer upon call of `next`.
+		///
+		/// E.g. One might uses this info in order to sequence a runtime upgrade or in order to
+		/// split a large upgrade of one pallet into multiple steps. So that each step fits into a
+		/// block.
+		type StateInfo: sp_std::default::Default + FullCodec;
+
+		/// General runtime upgrade info.
+		/// The `RuntimeVersion` and the `Memo` are stored under the block number the upgrade was
+		/// started.
+		fn info() -> (RuntimeVersion, Self::Memo);
+
+		/// The accumulated weight of the runtime upgrade. This value is used to compute the
+		/// number of blocks, that this upgrade will block the chain.
+		fn upgrade_weight() -> Weight;
+
+		/// The runtime calls the next part of the runtime to be run. Hereby, the runtime
+		/// indicates how much weight, the part must at most use.
+		///
+		/// It is up to the calley to ensure, that `usable` is not exceeded by the upgrade!
+		fn next(
```

Comment:
woudln't it better to let this migration pallet decide on how to batch the txns over multiple blocks instad of delegating the logic of batching to another implemntation ?
or do you think we may to version the impls of `RuntimeUpgradeProvider` ?

Reply:
Yes, I wanted to version the impls. That every upgrade has its own impl and even can have different `StateInfos` coming back to it, as we don't know what they will need in order to keep a state from one block to another.